File: python-legacy/src/elbencho_harness/report/charts.py
"""Plotly figure factories with a polished dark theme + interactive Bar/Line toggle.

Public API:
  - headline_tiles(result)                          one-result tile dicts
  - throughput_per_phase([results])                 single-spec phase bars
  - latency_overview([results])                     single-spec lat min/avg/max
  - sweep_overlay(...)                              multi-series overlay
                                                    with Bar/Line toggle
  - render_figure_html(fig)                         render to plain HTML

Design notes:
- ``_apply_theme`` puts every figure on the same dark visual language as the
  page (transparent backgrounds, soft gridlines, GitHub-style palette).
- ``_format_axis_value`` formats bytes ("64k", "1m"), counts ("8"), percents
  ("70%"), so x-tick labels stay terse.
- ``_detect_common_axis`` decides whether the sweep is one-dimensional (in
  which case the x-axis can be a clean numeric scale labeled by axis value)
  or mixed (fall back to per-point composite labels).
"""

from __future__ import annotations

from typing import Iterable, Sequence

import plotly.graph_objects as go

from ..results.schema import PhaseResult, Result


# ---------------------------------------------------------------------------
# Theme + colors
# ---------------------------------------------------------------------------

# GitHub-style dark palette, accessible (WCAG AA on #0e1116 background).
PALETTE = [
    "#58a6ff",  # blue
    "#3fb950",  # green
    "#d29922",  # amber
    "#f85149",  # red
    "#bc8cff",  # purple
    "#ff7b72",  # coral
    "#79c0ff",  # light blue
    "#56d364",  # light green
]

_FONT_FAMILY = (
    "-apple-system, BlinkMacSystemFont, 'Segoe UI', system-ui, "
    "Roboto, 'Helvetica Neue', Arial, sans-serif"
)
_GRID = "rgba(255,255,255,0.06)"
_AXIS = "rgba(255,255,255,0.18)"
_FG = "#e6edf3"
_MUTED = "#8b949e"
_PANEL = "#161b22"


def _apply_theme(
    fig: go.Figure,
    *,
    title: str | None = None,
    x_title: str | None = None,
    y_title: str | None = None,
    height: int = 460,
) -> None:
    """Apply the project's dark theme in place."""
    fig.update_layout(
        title=dict(
            text=title or "",
            x=0.02,
            y=0.97,
            font=dict(size=15, color=_FG, family=_FONT_FAMILY, weight=600),
        ),
        plot_bgcolor="rgba(0,0,0,0)",
        paper_bgcolor="rgba(0,0,0,0)",
        font=dict(family=_FONT_FAMILY, size=12, color=_FG),
        height=height,
        margin=dict(l=60, r=24, t=70, b=60),
        colorway=PALETTE,
        legend=dict(
            orientation="h",
            yanchor="bottom",
            y=1.02,
            xanchor="left",
            x=0.02,
            bgcolor="rgba(0,0,0,0)",
            font=dict(size=12, color=_FG),
        ),
        hovermode="x unified",
        hoverlabel=dict(
            bgcolor=_PANEL,
            bordercolor=_AXIS,
            font=dict(color=_FG, family=_FONT_FAMILY, size=12),
        ),
    )
    fig.update_xaxes(
        title=dict(text=x_title or "", font=dict(size=12, color=_MUTED)),
        gridcolor=_GRID,
        zerolinecolor=_GRID,
        linecolor=_AXIS,
        tickcolor=_AXIS,
        tickfont=dict(color=_FG, size=12),
        ticks="outside",
    )
    fig.update_yaxes(
        title=dict(text=y_title or "", font=dict(size=12, color=_MUTED)),
        gridcolor=_GRID,
        zerolinecolor=_GRID,
        linecolor=_AXIS,
        tickcolor=_AXIS,
        tickfont=dict(color=_FG, size=12),
        ticks="outside",
        rangemode="tozero",
    )


def overlay_with_toggle(
    fig: go.Figure, *, n_series: int, initial_mode: str = "bar"
) -> str:
    """Render a sweep_overlay figure wrapped in our custom Bar/Line toggle.

    Plotly's updatemenus buttons render with an active-state background that
    Plotly won't let you fully style; the "active" button ends up unreadable
    on a dark theme. So we strip the updatemenus and put a styled HTML
    button-group on top of the chart that calls Plotly.restyle() directly.

    The template (compare.html.j2 etc.) provides the matching CSS + JS handler
    once, so each chart only needs this wrapper.
    """
    chart_html = render_figure_html(fig)
    bar_active = "active" if initial_mode == "bar" else ""
    line_active = "active" if initial_mode == "line" else ""
    return (
        f'<div class="chart-wrapper" data-nseries="{n_series}">'
        f'  <div class="chart-toggle" role="group" aria-label="Chart type">'
        f'    <button type="button" data-mode="bar" class="{bar_active}"'
        f'            onclick="elbenchoToggleChart(this)">Bar</button>'
        f'    <button type="button" data-mode="line" class="{line_active}"'
        f'            onclick="elbenchoToggleChart(this)">Line</button>'
        f'  </div>'
        f'  {chart_html}'
        f'</div>'
    )


# ---------------------------------------------------------------------------
# Value formatting
# ---------------------------------------------------------------------------


def _human_bytes(n: int) -> str:
    """65536 -> '64k', 1048576 -> '1m', 1073741824 -> '1g'. Lowercase.

    Uses binary units (1024-based) without the 'i' suffix to keep labels terse.
    Sub-1k values render with a 'b' suffix so the axis stays uniform.
    """
    n = int(n)
    for unit, base in (("g", 1024 ** 3), ("m", 1024 ** 2), ("k", 1024)):
        if n >= base and n % base == 0:
            return f"{n // base}{unit}"
        if n >= base:
            v = n / base
            return f"{v:g}{unit}"
    return f"{n}b" if n < 1024 else str(n)


def format_axis_value(axis: str, value) -> str:
    """Format one sweep-axis value for an x-tick label.

    block_size / dataset_size: bytes ('64k', '4m')
    rw_mix_pct_read:           percent ('70%')
    everything else:           str(value) ('8', '16')
    """
    if value is None:
        return ""
    if axis in {"block_size", "dataset_size"}:
        return _human_bytes(value)
    if axis == "rw_mix_pct_read":
        return f"{int(value)}%"
    return str(value)


def axis_pretty_name(axis: str) -> str:
    return {
        "block_size": "block size",
        "rw_mix_pct_read": "read mix",
        "threads_per_client": "threads / client",
        "io_depth": "IO depth",
        "dataset_size": "dataset size",
        "client_count": "clients",
    }.get(axis, axis)


def detect_common_axis(axes_dicts: Sequence[dict | None]) -> str | None:
    """If every dict has exactly one key and they're all the same key, return it.

    Otherwise None (caller falls back to composite labels).
    """
    axis_names: set[str] = set()
    for d in axes_dicts:
        if not d:
            return None
        keys = list(d.keys())
        if len(keys) != 1:
            return None
        axis_names.add(keys[0])
    if len(axis_names) == 1:
        return axis_names.pop()
    return None


# ---------------------------------------------------------------------------
# Single-spec charts (used by render_single)
# ---------------------------------------------------------------------------


def _safe(v):
    return v if v is not None else 0


def headline_tiles(result: Result) -> list[dict]:
    """Return a list of {label, value, unit, fmt, subtle} dicts for HTML cards.

    Pulls from the primary phase (read for 100% read, write for 100% write,
    mixed for everything else).
    """
    phase = result.phases.get(result.primary_phase)
    if phase is None:
        phase = next(iter(result.phases.values()), None)
    tiles: list[dict] = []
    if phase is None:
        return tiles

    throughput = phase.throughput_mib_s_last or phase.throughput_mib_s_first
    iops = phase.iops_last or phase.iops_first
    ops = phase.ops_per_s_last or phase.ops_per_s_first
    lat_avg = phase.io_lat_us.avg if phase.io_lat_us else None
    lat_max = phase.io_lat_us.max if phase.io_lat_us else None

    tiles.append({"label": "Throughput", "value": throughput, "unit": "MiB/s", "fmt": ",.1f"})
    if iops is not None:
        tiles.append({"label": "IOPS", "value": iops, "unit": "", "fmt": ",.0f"})
    if ops is not None:
        tiles.append({"label": "Ops/s", "value": ops, "unit": "", "fmt": ",.0f"})
    tiles.append({"label": "Avg latency", "value": lat_avg, "unit": "µs", "fmt": ",.1f"})
    tiles.append({"label": "Max latency", "value": lat_max, "unit": "µs", "fmt": ",.1f"})
    tiles.append({"label": "CPU", "value": phase.cpu_pct, "unit": "%", "fmt": ".1f"})
    tiles.append(
        {"label": "Errors", "value": phase.errors, "unit": "", "fmt": "d",
         "subtle": phase.errors == 0}
    )
    tiles.append(
        {"label": "Duration", "value": result.duration_s, "unit": "s", "fmt": ".1f",
         "subtle": True}
    )
    return tiles


def throughput_per_phase(results: Iterable[Result]) -> go.Figure:
    """Bar chart: one bar per (result, phase) showing throughput MiB/s."""
    fig = go.Figure()
    phase_order = ["write", "read", "mixed"]
    seen_phases: set[str] = set()
    for res in results:
        for phase_name in phase_order:
            if phase_name not in res.phases:
                continue
            seen_phases.add(phase_name)
        for phase_name, _ in res.phases.items():
            if phase_name not in phase_order:
                seen_phases.add(phase_name)
    ordered_phases = [p for p in phase_order if p in seen_phases] + sorted(
        seen_phases - set(phase_order)
    )

    for i, res in enumerate(results):
        label = f"{res.target.name}·{res.workload.name}"
        ys = []
        for phase_name in ordered_phases:
            phase = res.phases.get(phase_name)
            ys.append(_safe(phase.throughput_mib_s_last if phase else None) or
                      _safe(phase.throughput_mib_s_first if phase else None))
        fig.add_bar(name=label, x=ordered_phases, y=ys,
                    marker_color=PALETTE[i % len(PALETTE)], hovertemplate="%{y:,.1f} MiB/s<extra>%{fullData.name}</extra>")
    _apply_theme(fig, title="Throughput by phase", y_title="MiB/s", x_title="phase")
    fig.update_layout(barmode="group", bargap=0.25)
    return fig


def latency_overview(results: Iterable[Result]) -> go.Figure | None:
    """min / avg / max latency per phase as grouped bars."""
    xs: list[str] = []
    mins: list[float] = []
    avgs: list[float] = []
    maxs: list[float] = []
    for res in results:
        for phase_name, phase in res.phases.items():
            xs.append(phase_name)
            mins.append(_safe(phase.io_lat_us.min))
            avgs.append(_safe(phase.io_lat_us.avg))
            maxs.append(_safe(phase.io_lat_us.max))
    if not xs:
        return None
    fig = go.Figure()
    fig.add_bar(name="min", x=xs, y=mins, marker_color=PALETTE[1],
                hovertemplate="%{y:,.0f} µs<extra>min</extra>")
    fig.add_bar(name="avg", x=xs, y=avgs, marker_color=PALETTE[0],
                hovertemplate="%{y:,.0f} µs<extra>avg</extra>")
    fig.add_bar(name="max", x=xs, y=maxs, marker_color=PALETTE[3],
                hovertemplate="%{y:,.0f} µs<extra>max</extra>")
    _apply_theme(fig, title="IO latency (min / avg / max)", y_title="µs", x_title="phase")
    fig.update_layout(barmode="group", bargap=0.25)
    return fig


# ---------------------------------------------------------------------------
# Sweep overlay (compare report)
# ---------------------------------------------------------------------------


def sweep_overlay(
    *,
    title: str,
    x_labels: list[str],
    x_axis_title: str,
    y_title: str,
    series: list[tuple[str, list[float | None]]],
    hover_unit: str = "",
    initial_mode: str = "bar",
    sort_numeric: bool = False,
) -> go.Figure:
    """Multi-series chart with a live Bar/Line toggle (Plotly updatemenus).

    Both representations (bars + lines) are baked in at render time; the
    toggle flips trace visibility on the client. No JS callbacks needed.

    Args:
      title:         chart title
      x_labels:      one x value per data point (already formatted)
      x_axis_title:  axis label (e.g. "block size")
      y_title:       y-axis label
      series:        list of (run_label, [y values aligned with x_labels])
      hover_unit:    appended to hover values, e.g. "MiB/s"
      initial_mode:  "bar" or "line"
      sort_numeric:  if x_labels look like sized values (k/m/g suffixes or pure
                     ints), keep them ordered as given but display on a
                     numeric axis. Otherwise treat as categorical.
    """
    fig = go.Figure()
    n = len(series)
    # Bar traces first (initially visible).
    for i, (name, ys) in enumerate(series):
        color = PALETTE[i % len(PALETTE)]
        fig.add_bar(
            name=name,
            x=x_labels,
            y=[_safe(v) for v in ys],
            visible=(initial_mode == "bar"),
            marker_color=color,
            marker_line_width=0,
            hovertemplate=(
                "%{x} • %{y:,.1f} " + hover_unit + "<extra>" + name + "</extra>"
            ).strip(),
        )
    # Scatter+line traces second (initially hidden). Visibility flips the
    # legend entry too, so we get one legend entry per run in either mode.
    for i, (name, ys) in enumerate(series):
        color = PALETTE[i % len(PALETTE)]
        fig.add_scatter(
            name=name,
            x=x_labels,
            y=[_safe(v) for v in ys],
            mode="lines+markers",
            visible=(initial_mode == "line"),
            line=dict(color=color, width=2.5, shape="spline", smoothing=0.6),
            marker=dict(color=color, size=8, line=dict(width=2, color="#0e1116")),
            hovertemplate=(
                "%{x} • %{y:,.1f} " + hover_unit + "<extra>" + name + "</extra>"
            ).strip(),
        )

    _apply_theme(fig, title=title, x_title=x_axis_title, y_title=y_title)
    fig.update_layout(barmode="group", bargap=0.25)
    # The Bar/Line toggle lives in HTML alongside the chart (see
    # overlay_with_toggle) so it can be styled to match the rest of the page.
    # The figure carries both trace types; toggle flips trace visibility via
    # Plotly.restyle().
    return fig


# ---------------------------------------------------------------------------
# Rendering
# ---------------------------------------------------------------------------


def render_figure_html(fig: go.Figure | None, *, full_html: bool = False) -> str:
    if fig is None:
        return ""
    return fig.to_html(
        include_plotlyjs=False,
        full_html=full_html,
        config={"displaylogo": False, "responsive": True, "displayModeBar": False},
    )

File: src/report/charts.rs
//! Plotly.js JSON figure builders.
//!
//! Each function returns a JSON Value representing a complete Plotly figure
//! ({"data": [...], "layout": {...}}). The HTML templates embed these as
//! <script type="application/json"> and mount via Plotly.newPlot in the browser.
//!
//! Theme matches the Python version (GitHub dark palette).

use serde_json::{json, Value};

use crate::results::schema::Result as RunResult;

pub const PALETTE: &[&str] = &[
    "#58a6ff", "#3fb950", "#d29922", "#f85149", "#bc8cff", "#ff7b72", "#79c0ff", "#56d364",
];

const FONT_FAMILY: &str =
    "-apple-system, BlinkMacSystemFont, 'Segoe UI', system-ui, Roboto, 'Helvetica Neue', Arial, sans-serif";
const GRID: &str = "rgba(255,255,255,0.06)";
const AXIS: &str = "rgba(255,255,255,0.18)";
const FG: &str = "#e6edf3";
const MUTED: &str = "#8b949e";
const PANEL: &str = "#161b22";

/// Build a layout object with our standard dark theme.
fn theme_layout(title: &str, x_title: &str, y_title: &str, height: u32) -> Value {
    json!({
        "title": {
            "text": title,
            "x": 0.02,
            "y": 0.97,
            "font": {"size": 15, "color": FG, "family": FONT_FAMILY}
        },
        "plot_bgcolor": "rgba(0,0,0,0)",
        "paper_bgcolor": "rgba(0,0,0,0)",
        "font": {"family": FONT_FAMILY, "size": 12, "color": FG},
        "height": height,
        "margin": {"l": 60, "r": 24, "t": 70, "b": 60},
        "colorway": PALETTE,
        "legend": {
            "orientation": "h",
            "yanchor": "bottom",
            "y": 1.02,
            "xanchor": "left",
            "x": 0.02,
            "bgcolor": "rgba(0,0,0,0)",
            "font": {"size": 12, "color": FG}
        },
        "hovermode": "x unified",
        "hoverlabel": {
            "bgcolor": PANEL,
            "bordercolor": AXIS,
            "font": {"color": FG, "family": FONT_FAMILY, "size": 12}
        },
        "xaxis": {
            "title": {"text": x_title, "font": {"size": 12, "color": MUTED}},
            "gridcolor": GRID,
            "zerolinecolor": GRID,
            "linecolor": AXIS,
            "tickcolor": AXIS,
            "tickfont": {"color": FG, "size": 12},
            "ticks": "outside"
        },
        "yaxis": {
            "title": {"text": y_title, "font": {"size": 12, "color": MUTED}},
            "gridcolor": GRID,
            "zerolinecolor": GRID,
            "linecolor": AXIS,
            "tickcolor": AXIS,
            "tickfont": {"color": FG, "size": 12},
            "ticks": "outside",
            "rangemode": "tozero"
        },
        "barmode": "group",
        "bargap": 0.25
    })
}

/// Throughput chart for a single-spec report. Bars per phase.
pub fn single_throughput(result: &RunResult) -> Value {
    let phase_order = ["write", "read", "mixed"];
    let mut xs: Vec<&str> = Vec::new();
    let mut ys: Vec<f64> = Vec::new();
    for &name in &phase_order {
        if let Some(p) = result.phases.get(name) {
            xs.push(name);
            ys.push(p.throughput_mib_s_last.or(p.throughput_mib_s_first).unwrap_or(0.0));
        }
    }
    let label = format!("{}·{}", result.target.name, result.workload.name);
    let data = vec![json!({
        "type": "bar",
        "name": label,
        "x": xs,
        "y": ys,
        "marker": {"color": PALETTE[0]},
        "hovertemplate": "%{y:,.1f} MiB/s<extra>%{fullData.name}</extra>"
    })];
    json!({"data": data, "layout": theme_layout("Throughput by phase", "phase", "MiB/s", 420)})
}

/// Latency chart for a single-spec report. Min/avg/max bars per phase.
pub fn single_latency(result: &RunResult) -> Option<Value> {
    let mut xs: Vec<&str> = Vec::new();
    let mut mins: Vec<f64> = Vec::new();
    let mut avgs: Vec<f64> = Vec::new();
    let mut maxs: Vec<f64> = Vec::new();
    for (name, p) in &result.phases {
        xs.push(name);
        mins.push(p.io_lat_us.min.unwrap_or(0.0));
        avgs.push(p.io_lat_us.avg.unwrap_or(0.0));
        maxs.push(p.io_lat_us.max.unwrap_or(0.0));
    }
    if xs.is_empty() {
        return None;
    }
    let data = vec![
        json!({
            "type": "bar", "name": "min", "x": xs, "y": mins,
            "marker": {"color": PALETTE[1]},
            "hovertemplate": "%{y:,.0f} µs<extra>min</extra>"
        }),
        json!({
            "type": "bar", "name": "avg", "x": xs, "y": avgs,
            "marker": {"color": PALETTE[0]},
            "hovertemplate": "%{y:,.0f} µs<extra>avg</extra>"
        }),
        json!({
            "type": "bar", "name": "max", "x": xs, "y": maxs,
            "marker": {"color": PALETTE[3]},
            "hovertemplate": "%{y:,.0f} µs<extra>max</extra>"
        }),
    ];
    Some(json!({"data": data, "layout": theme_layout("IO latency (min / avg / max)", "phase", "µs", 420)}))
}

/// Multi-run overlay chart with both bar AND scatter+line trace pairs baked
/// in. Initial mode is bar (visible); the lines are hidden. The compare HTML
/// template's elbenchoToggleChart() flips visibility on click.
pub fn overlay(
    title: &str,
    x_axis_title: &str,
    y_title: &str,
    hover_unit: &str,
    x_labels: &[String],
    series: &[(String, Vec<Option<f64>>)],
) -> Value {
    let mut data: Vec<Value> = Vec::with_capacity(series.len() * 2);
    // Bars first.
    for (i, (name, ys)) in series.iter().enumerate() {
        let color = PALETTE[i % PALETTE.len()];
        let yvals: Vec<f64> = ys.iter().map(|v| v.unwrap_or(0.0)).collect();
        data.push(json!({
            "type": "bar",
            "name": name,
            "x": x_labels,
            "y": yvals,
            "visible": true,
            "marker": {"color": color, "line": {"width": 0}},
            "hovertemplate": format!(
                "%{{x}} • %{{y:,.1f}} {}<extra>{}</extra>", hover_unit, name
            )
        }));
    }
    // Scatter lines second (hidden initially).
    for (i, (name, ys)) in series.iter().enumerate() {
        let color = PALETTE[i % PALETTE.len()];
        let yvals: Vec<f64> = ys.iter().map(|v| v.unwrap_or(0.0)).collect();
        data.push(json!({
            "type": "scatter",
            "name": name,
            "x": x_labels,
            "y": yvals,
            "mode": "lines+markers",
            "visible": false,
            "line": {"color": color, "width": 2.5, "shape": "spline", "smoothing": 0.6},
            "marker": {"color": color, "size": 8, "line": {"width": 2, "color": "#0e1116"}},
            "hovertemplate": format!(
                "%{{x}} • %{{y:,.1f}} {}<extra>{}</extra>", hover_unit, name
            )
        }));
    }
    json!({"data": data, "layout": theme_layout(title, x_axis_title, y_title, 460)})
}

/// Format an axis value for chart x-tick labels.
///
/// block_size / dataset_size: bytes -> "64k", "1m", "4m" (lowercase)
/// rw_mix_pct_read:           "70%"
/// integers (threads, qd, ...): "8"
pub fn format_axis_value(axis: &str, value: i64) -> String {
    match axis {
        "block_size" | "dataset_size" => human_bytes(value as u64),
        "rw_mix_pct_read" => format!("{}%", value),
        _ => format!("{}", value),
    }
}

fn human_bytes(n: u64) -> String {
    for (unit, base) in [("g", 1u64 << 30), ("m", 1u64 << 20), ("k", 1u64 << 10)] {
        if n >= base && n % base == 0 {
            return format!("{}{}", n / base, unit);
        }
        if n >= base {
            return format!("{}{}", (n as f64) / (base as f64), unit);
        }
    }
    if n >= 1024 {
        format!("{}", n)
    } else {
        format!("{}b", n)
    }
}

pub fn axis_pretty_name(axis: &str) -> &'static str {
    match axis {
        "block_size" => "block size",
        "rw_mix_pct_read" => "read mix",
        "threads_per_client" => "threads / client",
        "io_depth" => "IO depth",
        "dataset_size" => "dataset size",
        "client_count" => "clients",
        _ => "",
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn human_bytes_canonical() {
        assert_eq!(human_bytes(65_536), "64k");
        assert_eq!(human_bytes(1_048_576), "1m");
        assert_eq!(human_bytes(4 * 1_048_576), "4m");
        assert_eq!(human_bytes(1 << 30), "1g");
    }

    #[test]
    fn axis_value_formatting() {
        assert_eq!(format_axis_value("block_size", 65_536), "64k");
        assert_eq!(format_axis_value("threads_per_client", 8), "8");
        assert_eq!(format_axis_value("rw_mix_pct_read", 70), "70%");
    }

    #[test]
    fn overlay_includes_bar_and_scatter_pairs() {
        let fig = overlay(
            "T",
            "x",
            "MiB/s",
            "MiB/s",
            &["64k".into(), "1m".into()],
            &[
                ("run-A".into(), vec![Some(100.0), Some(200.0)]),
                ("run-B".into(), vec![Some(110.0), Some(210.0)]),
            ],
        );
        let data = fig["data"].as_array().unwrap();
        assert_eq!(data.len(), 4);
        let bars: Vec<_> = data.iter().filter(|t| t["type"] == "bar").collect();
        let scatters: Vec<_> = data.iter().filter(|t| t["type"] == "scatter").collect();
        assert_eq!(bars.len(), 2);
        assert_eq!(scatters.len(), 2);
        // Bars visible, scatters hidden initially.
        assert_eq!(bars[0]["visible"], serde_json::Value::Bool(true));
        assert_eq!(scatters[0]["visible"], serde_json::Value::Bool(false));
    }
}
